jax/_src/lax/search.py

```diff
@@ -0,0 +1,179 @@
+  if not batch_dims <= dimension < sorted_arr.ndim:
+    raise ValueError(f"dimension={dimension} must be in range [{batch_dims}, {sorted_arr.ndim})")
+  if sorted_arr.shape[:batch_dims] != query.shape[:batch_dims]:
+    raise ValueError(f"batch dimension sizes must match; got {sorted_arr.shape[:batch_dims]} != "
+                     f"{query.shape[:batch_dims]}")
+  shape = (*sorted_arr.shape[:batch_dims],
+           *(s for d, s in enumerate(sorted_arr.shape) if d >= batch_dims and d != dimension),
+           *(s for d, s in enumerate(query.shape) if d >= batch_dims))
+  dtype = np.int32 if sorted_arr.shape[dimension] < np.iinfo(np.int32).max else np.int64
+  return core.ShapedArray(shape, dtype)
+
+def _searchsorted_impl(sorted_arr, query, *, side, dimension, batch_dims, method):
+  if method == "scan":
+    _searchsorted = _searchsorted_scan_unbatched
+  elif method == "sort":
+    _searchsorted = _searchsorted_sort_unbatched
+  else:
+    raise ValueError(f"invalid argument method={method!r}; expected 'sort' or 'scan'")
+  out_aval = _searchsorted_abstract_eval(sorted_arr, query, side=side, dimension=dimension,
+                                         batch_dims=batch_dims, method=method)
+  sorted_arr = batching.moveaxis(sorted_arr, dimension, -1)
+  fun = partial(_searchsorted, side=side, dtype=out_aval.dtype)
+  for _ in range(sorted_arr.ndim - batch_dims - 1):
+    fun = api.vmap(fun, in_axes=(0, None))
+  for _ in range(batch_dims):
+    fun = api.vmap(fun, in_axes=0)
+  return fun(sorted_arr, query)
+
+@partial(jax.jit, static_argnames=['side', 'dtype'])
+def _searchsorted_scan_unbatched(sorted_arr, query, *, side, dtype):
+  assert sorted_arr.ndim == 1
+  assert side in ['left', 'right']
+  if len(sorted_arr) == 0:
+    return lax._zeros(query, dtype=dtype)
+  if query.ndim > 0:
+    return api.vmap(partial(_searchsorted_scan_unbatched, side=side, dtype=dtype), in_axes=(None, 0))(sorted_arr, query)
+
+  op = lax._sort_le_comparator if side == 'left' else lax._sort_lt_comparator
+
+  def body_fun(i, state):
+    low, high = state
+    mid = (low + high) // 2
+    go_left = op(query, sorted_arr[mid])
+    return (lax.select(go_left, low, mid), lax.select(go_left, mid, high))
+
+  N, = sorted_arr.shape
+  n_levels = int(np.ceil(np.log2(N + 1)))
+  return fori_loop(0, n_levels, body_fun, (dtype.type(0), dtype.type(N)))[1]
+
+@partial(jax.jit, static_argnames=['side', 'dtype'])
+def _searchsorted_sort_unbatched(sorted_arr, query, *, side, dtype):
+  assert sorted_arr.ndim == 1
+  assert side in ['left', 'right']
+  def _rank(x):
+    idx = lax.iota(dtype, len(x))
+    return lax._zeros(idx).at[lax.sort_key_val(x, idx)[1]].set(idx)
+  if side == 'left':
+    index = _rank(lax.concatenate([query.ravel(), sorted_arr], 0))[:query.size]
+  else:
+    index = _rank(lax.concatenate([sorted_arr, query.ravel()], 0))[sorted_arr.size:]
+  return lax.reshape(lax.sub(index, _rank(index)), np.shape(query))
```

Comment:
Will changing `_rank(index)` into `_rank(query)` be helpful, as then `index` and `_rank(query)` can be scheduled in parallel (with data dependency removed)?

Reply:
Good call! Done.